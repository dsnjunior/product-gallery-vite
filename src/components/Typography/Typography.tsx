import clsx from 'clsx';

export type TypographyProps = {
  variant?: 'h600' | 'h500' | 'h400' | 'h300' | 'h200' | 'h100' | 'p300' | 'p200' | 'c200' | 'c100' | 'l100';
  weight?: '400' | '500' | '600' | '700';
  transform?: 'uppercase' | 'lowercase' | 'capitalize';
  className?: string;
  component?: React.ElementType;
  children: React.ReactNode;
  title?: true | string;
};

const elements = {
  h600: 'h1',
  h500: 'h2',
  h400: 'h3',
  h300: 'h4',
  h200: 'h5',
  h100: 'h6',
  p300: 'p',
  p200: 'p',
  c200: 'p',
  c100: 'p',
  l100: 'label',
} as const;

const variants = {
  h600: 'text-h600',
  h500: 'text-h500',
  h400: 'text-h400',
  h300: 'text-h300',
  h200: 'text-h200',
  h100: 'text-h100',
  p300: 'text-p300',
  p200: 'text-p200',
  c200: 'text-c200',
  c100: 'text-c100',
  l100: 'text-l100',
} as const;

const weights = {
  400: 'font-400',
  500: 'font-500',
  600: 'font-600',
  700: 'font-700',
} as const;

const transforms = {
  uppercase: 'uppercase',
  lowercase: 'lowercase',
  capitalize: 'capitalize',
} as const;

export function Typography({
  variant = 'p300',
  weight,
  transform,
  className,
  children,
  component,
  title,
}: TypographyProps) {
  const Element = component ?? elements[variant];

  return (
    <Element
      className={clsx(
        variants[variant],
        weight ? weights[weight] : '',
        transform ? transforms[transform] : '',
        className,
      )}
      title={typeof title === 'string' || typeof title === 'undefined' ? title : children}
    >
      {children}
    </Element>
  );
}
