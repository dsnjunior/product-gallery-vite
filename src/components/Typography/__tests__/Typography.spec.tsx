import { render, screen } from '@testing-library/react';
import { describe, it } from 'vitest';

import { Typography } from '../Typography';

describe('Typography.tsx', () => {
  it('should match snapshot', async () => {
    render(<Typography>Test</Typography>);

    expect(await screen.findByText('Test')).toMatchSnapshot();
  });

  it('should render children', async () => {
    render(<Typography>Test</Typography>);

    expect(await screen.findByText('Test')).toBeInTheDocument();
  });

  it.each([
    { variant: 'h600' as const, tag: 'h1' },
    { variant: 'h500' as const, tag: 'h2' },
    { variant: 'h400' as const, tag: 'h3' },
    { variant: 'h300' as const, tag: 'h4' },
    { variant: 'h200' as const, tag: 'h5' },
    { variant: 'h100' as const, tag: 'h6' },
    { variant: 'p300' as const, tag: 'p' },
    { variant: 'p200' as const, tag: 'p' },
    { variant: 'c200' as const, tag: 'p' },
    { variant: 'c100' as const, tag: 'p' },
    { variant: 'l100' as const, tag: 'label' },
  ])('variant $variant should have correct tag and className', async ({ variant, tag }) => {
    render(<Typography variant={variant}>Test</Typography>);

    const element = await screen.findByText('Test');
    expect(element.tagName.toLocaleLowerCase()).toBe(tag);
    expect(element).toHaveClass(`text-${variant}`);
  });

  it('should render custom component', async () => {
    render(<Typography component="div">Test</Typography>);

    expect((await screen.findByText('Test')).tagName.toLocaleLowerCase()).toBe('div');
  });

  it('should render custom className', async () => {
    render(<Typography className="custom">Test</Typography>);

    expect(await screen.findByText('Test')).toHaveClass('custom');
  });

  it('should render custom weight', async () => {
    render(<Typography weight="600">Test</Typography>);

    expect(await screen.findByText('Test')).toHaveClass('font-600');
  });

  it('should render custom transform', async () => {
    render(<Typography transform="uppercase">Test</Typography>);

    expect(await screen.findByText('Test')).toHaveClass('uppercase');
  });

  it('should render custom weight and transform', async () => {
    render(
      <Typography weight="600" transform="uppercase">
        Test
      </Typography>,
    );

    const element = await screen.findByText('Test');
    expect(element).toHaveClass('font-600');
    expect(element).toHaveClass('uppercase');
  });

  it('should render custom weight and transform with variant', async () => {
    render(
      <Typography variant="h100" weight="600" transform="uppercase">
        Test
      </Typography>,
    );

    const element = await screen.findByText('Test');
    expect(element).toHaveClass('text-h100');
    expect(element).toHaveClass('font-600');
    expect(element).toHaveClass('uppercase');
  });
});
