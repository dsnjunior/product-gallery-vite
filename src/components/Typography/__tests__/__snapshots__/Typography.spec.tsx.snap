// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typography.tsx > should match snapshot 1`] = `
<p
  class="text-p300"
>
  Test
</p>
`;
